### PR TITLE
Add support flash driver for STM32F7x SoC

### PR DESCRIPTION
**Pull request by jinyongiot**
This patch add support flash driver for STM32F7x SoC. So the boards of STM32F7x SoC can use the flash_stm32 interface. Test on stm32f769i_disco

### CODE REVIEW

**Comment by aurel32** [comment]
I have also been working on flash support for STM32F7 during the week-end, and was about to submit it. My branch is available there: https://github.com/aurel32/zephyr/tree/stm32f7-flash . That way I know exactly what to look for, so I will review your changes.

File: drivers/flash/flash_stm32.c [hunk 1]
```diff
@@ -104,6 +107,14 @@ static void flash_stm32_flush_caches(struct device *dev)
...
 		regs->acr.val |= FLASH_ACR_DCEN;
 	}
+#else
+	if (regs->acr.val & FLASH_ACR_ARTEN) {
+		regs->acr.val &= ~FLASH_ACR_ARTEN;
+		regs->acr.val |= FLASH_ACR_ARTRST;
+		regs->acr.val &= ~FLASH_ACR_ARTRST;
+		regs->acr.val |= FLASH_ACR_ARTEN;
+	}
+#endif
```
Comment (aurel32) [inline drivers/flash/flash_stm32.c]: This is only valid for access through the ITCM bus. Usually the code is executed through the AXIS bus, and goes through the cache. The data cache should be invalidated through `SCB_InvalidateDCache_by_Addr`. I don't think we should flush the instruction cache after an erase. See https://github.com/aurel32/zephyr/commit/0188becc6d3e91933b6536509a49fe9953fe2acd

File: drivers/flash/flash_stm32f7x.c [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->cr &= ~CR_PSIZE_MASK;
+	regs->cr |= FLASH_PSIZE_BYTE;
+	regs->cr |= FLASH_CR_PG;
+
+	/* flush the register write */
+	tmp = regs->cr;
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: The reference manual recommends to use a DSB instruction to flush the register. The Cortex-M7 is using a more advanced pipeline, so the trick of doing a dummy read might not be enough.

File: drivers/flash/flash_stm32f7x.c [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+	*((u8_t *) offset + CONFIG_FLASH_BASE_ADDRESS) = val;
+
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: You should insert a DSB instruction as recommended in the reference manual. I actually have seen issues on the STM32F723 where some data when not written to the flash.

File: drivers/flash/flash_stm32f7x.c [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		return rc;
+	}
+
+	regs->cr &= STM32F7X_SECTOR_MASK;
+	regs->cr |= FLASH_CR_SER | (sector << 3);
+	regs->cr |= FLASH_CR_STRT;
+
+	/* flush the register write */
+	tmp = regs->cr;
+
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: Same comment about adding a DSB instruction here.

File: drivers/flash/flash_stm32f7x.c [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+ * reference manuals, but the flash layout for a given number of
+ * sectors is consistent across these manuals, with one "gotcha". The
+ * number of sectors is given by the HAL as FLASH_SECTOR_TOTAL.
+ *
+ * The only "gotcha" is that when there are 24 sectors, they are split
+ * across 2 "banks" of 12 sectors each, with another set of small
+ * sectors (16 KB) in the second bank occurring after the large ones
+ * (128 KB) in the first. We could consider supporting this as two
+ * devices to make the layout cleaner, but this will do for now.
+ */
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: That is not true for the STM32F769 which can be configured in two different way, with either 12 sectors (4*32K, 128K, 7*256K) or 24 sectors (twice 4*16K, 64K, 7*128K).

File: drivers/flash/flash_stm32f7x.c [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+#if FLASH_SECTOR_TOTAL == 8
+static const struct flash_pages_layout stm32f7_flash_layout[] = {
+	/*
+	 * RM0410, table 5
+	 */
+	{.pages_count = 4, .pages_size = KB(32)},
+	{.pages_count = 1, .pages_size = KB(128)},
+	{.pages_count = 3, .pages_size = KB(256)},
+};
+#elif FLASH_SECTOR_TOTAL == 12
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: Note that this case never appears in the STM32 HAL.

File: drivers/flash/flash_stm32f7x.c [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+static const struct flash_pages_layout stm32f7_flash_layout[] = {
+	/*
+	 * RM0410, table 3
+	 */
+	{.pages_count = 4, .pages_size = KB(32)},
+	{.pages_count = 1, .pages_size = KB(128)},
+	{.pages_count = 7, .pages_size = KB(256)},
+};
+#elif FLASH_SECTOR_TOTAL == 16
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: Same here, this case never appears in the STM32 HAL.

File: drivers/flash/flash_stm32f7x.c [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	/*
+	 * RM0410, table 4
+	 */
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+};
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: As said above, there are two cases depending of the `nDBANK` bit of the `FLASH_OPTCR` register. You could have a look how I have done that there: https://github.com/aurel32/zephyr/commit/7b6e90dd019dd68cee463c40230f23aa0fdbd81b#diff-322955a6f2ebc21dba22c2c642a8b3faR129

File: drivers/flash/flash_stm32f7x.c [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+#else
+#error "Unknown flash layout"
+#endif /* FLASH_SECTOR_TOTAL == 8 */
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: Could you please add the other STM32F7 SoCs? You could get that part there: https://github.com/aurel32/zephyr/commit/7b6e90dd019dd68cee463c40230f23aa0fdbd81b#diff-322955a6f2ebc21dba22c2c642a8b3faR129

**Review by aurel32 — CHANGES_REQUESTED** [review]
Could you please tell use what you tested with this flash driver? At least for me I was not able to do a sector erase without getting a timeout error. It can be quite long on the STM32F7 (up to 4 seconds for the biggest sectors). I have a patch for that there: 
https://github.com/aurel32/zephyr/commit/10d3541262fcc02e9922ba800db1237716afb873

**Review by superna9999 — CHANGES_REQUESTED** [review]
I'm ok with @aurel32 comments, please fix these issues and explicit how and on with SoC you did the tests

File: drivers/flash/flash_stm32f7x.c (hunk 9 above)
Comment (jinyongiot) [reply drivers/flash/flash_stm32f7x.c]: I am sorry, I don't test my code.
Thank you for pointing out the error of my code.
Now, I rewrite my code refer to your code, but test no pass on my board stm32f769i_disco too.
So, I want to write it refer to the STM32 HAL.

**Comment by jinyongiot** [comment]
Modify the code and test it time after time. It can run on the stm32f769i_disco board, now. The tested code path is samples / drivers / flash_shell.

File: drivers/flash/flash_stm32f7x.c [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+		return rc;
+	}
+	if (sector > 11) {
+		sector += 4;
+	}
+
+	regs->cr &= STM32F7X_SECTOR_MASK;
+	regs->cr |= FLASH_CR_SER | (sector << FLASH_CR_SNB_Pos);
+	regs->cr |= FLASH_CR_STRT;
+
```
Comment (aurel32) [inline drivers/flash/flash_stm32f7x.c]: Similarly to what is done for writing a byte, you should set the `PSIZE` field to `BYTE`. It's possible to get a bigger erase parallelism on the STM32F7, but it is not possible over the whole supported voltage range. Given we do not know the voltage, it should be forced to byte.

I believe this can be improved later by adding a new entry in the device tree, but that should be done later in an another patch.

File: drivers/flash/Kconfig.stm32 [hunk 11]
```diff
@@ -10,16 +10,18 @@ if FLASH && SOC_FAMILY_STM32
...
 	select FLASH_PAGE_LAYOUT if SOC_SERIES_STM32F4X
+	select FLASH_PAGE_LAYOUT if SOC_SERIES_STM32F7X
 	select FLASH_PAGE_LAYOUT if SOC_SERIES_STM32L4X
 	select FLASH_HAS_PAGE_LAYOUT if SOC_SERIES_STM32F0X
 	select FLASH_HAS_PAGE_LAYOUT if SOC_SERIES_STM32F4X
+	select FLASH_HAS_PAGE_LAYOUT if SOC_SERIES_STM32F7X
 	select FLASH_HAS_PAGE_LAYOUT if SOC_SERIES_STM32L4X
 	help
-	 Enable STM32F0x, STM32F3x, STM32F4x OR STM32L4x series flash driver.
+	 Enable STM32F0x, STM32F3x, STM32F4x STM32F7x OR STM32L4x series flash driver.
```
Comment (aurel32) [inline drivers/flash/Kconfig.stm32]: Minor things, there is a missing comma after STM32F4x

**Review by aurel32 — CHANGES_REQUESTED** [review]
Thanks for this new version, there are only minor things left. I have tried it on my STM32F723 discovery board using an NVS filesystem and confirm it all works fine.

File: drivers/flash/flash_stm32f7x.c [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+	if (sector > 11) {
+		sector += 4;
+	}
+
```
[the remark this answers is not in the document]
Comment (jinyongiot) [reply drivers/flash/flash_stm32f7x.c]: In dual bank mode, the SNB msb is selects the bank2, others select the sector in bank2.  The sector 12 is the sector 0 in bank2, so set the SNB to 0x10H (16D), this RM0410 page 110 say.

**Review by aurel32 — COMMENTED** [review]
Thanks for the update, most of the changes are fine, we are almost there. There are only a few minor things left around `STM32F7X_SECTOR_MASK`.

File: drivers/flash/flash_stm32f7x.c [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+#if FLASH_SECTOR_TOTAL == 24
+	/* Dual bank mode, SNB MSB selects the bank2, others select sector. */
+	if (sector > 11) {
```
[the remark this answers is not in the document]
Comment (jinyongiot) [reply drivers/flash/flash_stm32f7x.c]: By 1Mbytes devices, the sector8, sector9, sector10, sector11 are invalid. the first sector on bank2 is sector12, same as 2Mbytes devices. (table 6, RM0410)

File: drivers/flash/flash_stm32f7x.c [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+		sector += 4;
+	}
+#endif
+
+	/* clear PSIZE bits and SNB bits */
+	regs->cr &= STM32F7X_SECTOR_MASK;
+	regs->cr |= FLASH_PSIZE_BYTE;
```
[the remark this answers is not in the document]
Comment (jinyongiot) [reply drivers/flash/flash_stm32f7x.c]: you are right! 
And I like the first more.

File: drivers/flash/flash_stm32f7x.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (aurel32) [reply drivers/flash/flash_stm32f7x.c]: You are correct, I had the impression that we have to account for the missing sectors 7-11 for 1MB devices, but this will have to be done at the flash page layout level. Anyway those devices are not supported and the flash page layout API does not support defining holes in the map.

**Review by aurel32 — APPROVED** [review]
Thanks @jinyongiot for your work, from my point of view this is ready to be merged. @superna9999, do you still have some comments about this PR?

**Comment by aurel32** [comment]
recheck

**Comment by jinyongiot** [comment]
My suggestion: For stm32f7x 1Mbytes devices, we can use a sector number map to the devices sector number. Example, use sector8 map to the devices sector12 (first sector on the bank2).

File: drivers/flash/flash_stm32.c [hunk 15]
```diff
@@ -257,7 +272,7 @@ static int stm32_flash_init(struct device *dev)
 
 	k_sem_init(&p->sem, 1, 1);
 
-	return flash_stm32_write_protection(dev, false);
+	return flash_stm32_write_protection(dev, true);
```
Comment (superna9999) [inline drivers/flash/flash_stm32.c]: What is the reason to change this here ? Is this related with F7 support ?

Comment (jinyongiot) [reply drivers/flash/flash_stm32.c]: I just think that set flash protection false in init api is not necessary.

Comment (superna9999) [reply drivers/flash/flash_stm32.c]: Please push in a separate commit and a separate PR, this is not the right place to change that

**Review by superna9999 — APPROVED** [review]

File: drivers/flash/flash_stm32.c [hunk 16]
```diff
@@ -14,7 +14,19 @@
 
 #include "flash_stm32.h"
 
-#define STM32_FLASH_TIMEOUT	((u32_t) 0x000B0000)
+/* STM32F0: maximum erase time of 40ms for a 2K sector */
```
Comment (erwango) [inline drivers/flash/flash_stm32.c]: For better clarity, can you put the comments inside the ifdef's ?

File: drivers/flash/flash_stm32.c [hunk 17]
```diff
@@ -63,49 +75,50 @@ static int flash_stm32_check_status(struct device *dev)
...
 
 	if (regs->acr.val & FLASH_ACR_DCEN) {
 		regs->acr.val &= ~FLASH_ACR_DCEN;
 		regs->acr.val |= FLASH_ACR_DCRST;
 		regs->acr.val &= ~FLASH_ACR_DCRST;
 		regs->acr.val |= FLASH_ACR_DCEN;
 	}
-}
+#elif defined(CONFIG_SOC_SERIES_STM32F7X)
+	SCB_InvalidateDCache_by_Addr((void *)CONFIG_FLASH_BASE_ADDRESS
```
Comment (erwango) [inline drivers/flash/flash_stm32.c]: ARG_UNUSED(dev); ?

File: drivers/flash/flash_stm32f7x.c [hunk 18]
```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2017 Linaro Limited
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: Remove this one and put your own copyright here.

File: drivers/flash/flash_stm32f7x.c [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+
+#include <arch/cpu.h>
+#include <cortex_m/exc.h>
+
+#include "flash_stm32.h"
+
+#define STM32F7X_SECTOR_MASK		((u32_t) 0xFFFFFF07)
+
+bool flash_stm32_valid_range(struct device *dev, off_t offset, u32_t len,
+			     bool write)
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: Align 'bool' on 'struct' on at upper line

File: drivers/flash/flash_stm32f7x.c [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+	}
+/* Dual bank mode, SNB MSB selects the bank2,
+ * sothers select sector, So we remap sector number.
+ */
+#ifdef FLASH_OPTCR_nDBANK
+#if CONFIG_FLASH_SIZE == 1024
+	if (sector > 11) {
+		sector += 4;
+	}
+#elif CONFIG_FLASH_SIZE == 1024
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: You have same value in #if and #elif, is that intentional?

File: drivers/flash/flash_stm32f7x.c [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+	__DSB();
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	regs->cr &= ~(FLASH_CR_SER | FLASH_CR_SNB);
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(struct device *dev, unsigned int offset,
+				 unsigned int len)
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: align 'unsigned' on 'struct' in upper line

File: drivers/flash/flash_stm32f7x.c [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	return rc;
+}
+
+int flash_stm32_write_range(struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: Align 'const' on 'struct' in upper line

File: drivers/flash/flash_stm32f7x.c [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+		rc = write_byte(dev, offset, ((const u8_t *) data)[i]);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+/* Different SoC flash layouts are specified in across various
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: remove 'across'

File: drivers/flash/flash_stm32f7x.c [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+ * reference manuals, but the flash layout for a given number of
+ * sectors. In case of 8 sectors we need to differentiate between
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: 'but the flash layout for a given number of sectors.' Do we miss a word in this part of the sentence?

File: drivers/flash/flash_stm32f7x.c [hunk 25]
```diff
@@ -0,0 +1,249 @@
...
+	{.pages_count = 7, .pages_size = KB(128)},
+};
+#endif
+#else
+#error "Unknown flash layout"
+#endif/* !defined(FLASH_SECTOR_TOTAL) */
+
+#endif /* !defined(FLASH_OPTCR_nDBANK) */
+void flash_stm32_page_layout(struct device *dev,
+			     const struct flash_pages_layout **layout,
```
Comment (erwango) [inline drivers/flash/flash_stm32f7x.c]: Align 'const' on 'struct'

**Review by erwango — CHANGES_REQUESTED** [review]
Some comments but good work otherwise! Thanks

**Review by erwango — CHANGES_REQUESTED** [review]
Please take into account the comments about function header alignment.
Otherwise looks good

**Comment by erwango** [comment]
@jinyongiot , any update?

**Comment by erwango** [comment]
Closing as replaced by #12083
